+NeuroAnalysis/+IO/MetaTable.m

```diff
@@ -0,0 +1,128 @@
+            if exist(filepath, 'file')
+                metafile = load(filepath);
+                obj.Tests = metafile.Tests;
+            else
+                obj.Tests = struct([]);
+            end
+        end
+        
+        function addEntries(obj, tests, deleteMissing)
+            %ADDENTRIES Add multiple tests
+            % if deleteMissing is true, then any tests not supplied in the
+            % input will be removed
+            oldIdx = length(obj.Tests);
+            for t = 1:length(tests)
+                [obj.Tests, merged] = obj.merge(obj.Tests, tests{t});
+                oldIdx = oldIdx - merged;
+            end
+            if deleteMissing
+                obj.Tests = obj.Tests(oldIdx+1:end);
+            end
+        end
+        
+        function Tests = query(obj, varargin)
+            %QUERY Find matching tests
+            nargs = length(varargin);
+            if nargs == 0 || mod(nargs, 2) == 1
+               error('inputs should be name,value pairs, e.g. query(MetaTable, ''Subject_ID'', ''R1701'')')
+            end
+            keys = varargin(1:2:end);
+            values = varargin(2:2:end);
+            isMatch = true(length(obj.Tests),1);
+            for t = 1:length(obj.Tests)
+                for kv = 1:length(keys)
+                    isMatch(t) = isMatch(t) && isequal(obj.Tests(t).(keys{kv}), values{kv});
+                end
+            end
+            Tests = obj.Tests(isMatch);
+        end
+        
+        function Tests = list(obj)
+            %LIST list all the tests
+            Tests = obj.Tests;
+        end
+        
+        function export(obj, filepath)
+            %EXPORT save to disk
+            disp(['Saving metadata:    ',filepath,'    ...']);
+            metafile.Tests = obj.Tests;
+            fields = fieldnames(obj.Tests);
+            for f=1:length(fields)
+                metafile.TestsScalar.(fields{f}) = {obj.Tests.(fields{f})};
+            end
+            save(filepath, '-struct', 'metafile', '-v7.3');
+            disp('Saving metadata:    Done.');
+        end
+    end
+     
+    methods (Static, Access = private)
+
+         function [Tests, merged] = merge(Tests, test)
+            %MERGE Merge a test to a structure of tests
+
+            % Add missing fields to this object's Tests structure
+            fields = setdiff(fieldnames(test), fieldnames(Tests));
+            emptyCell = cell(length(Tests),1);
+            for f = 1:length(fields)
+                if any(strcmpi(fields{f}, fieldnames(Tests)))
+                    error(['Fieldnames must be unique and case-insensitive (%s)',...
+                        fields{f}]);
+                end
+                [Tests.(fields{f})] = emptyCell{:};
+            end
+            
+            % Merge according to sourceformat if necessary
+            toMerge = false(length(Tests),1);
+            for t = 1:length(Tests)
+                toMerge(t) = isequal(Tests(t).(test.key), test.(test.key));
```

Comment:
I'm still not sure about this; i've required that the metadata prepare function identify one field as a 'key' that uniquely identifies each test. For ripple data, this would be the filename, but for optical imaging / vlab data you would need to think of something else to use. Also I'm not sure if we need to be checking if the sourceformat matches as well, e.g. could there be multiple tests with the same filename but different source formats?

Reply:
i think the logical way is to first search for relevant entries by filtering on properties such as `sourceformat`, `test`, `subject`, etc. if result is empty means this is a new entry to be added, if result is one entry then merge, if greater than one, warn, record ambiguous entries and skip, we could deal with them later.